Keep expansion

garrigue opened the pull request:
This PR solves a number of long-standing problems in the OCaml typechecker
* Non-termination due to the creation of non-productive recursive types: #9314 
* Incorrect tracking of principality in presence of type abbreviations
* Incorrect detection of ambivalence in presence of type abbreviations (cf. new errors in `ambivalent-apply.ml`, which correspond to ambivalent types that were not properly detected)

Solving all these problems requires working on the fully expanded form of types, rather than the cascading expansions we have currently.
Namely, currently calling `expand_head` on a type node `ty` of the form `Tconstr(path,args,memo)` returns a new type node `ty'`, independent of `ty`, so that changes in level or scope on `ty'` are not reflected on `ty` (and reciprocally). Similarly `ty` and `ty'` are distinct for unification. The expansion `path -> ty'` is kept inside `memo`, but this is only intended for the on-the-fly creation of recursive types, and in particular `memo` is not kept during generalization/copying.

This PR adds a new kind of node descriptor, `Texpand(ty,path,args)`, which is essentially equivalent to `Tlink(ty)`, but also keeps the `path` that was expanded. This is infromation is used in printing, but appears to be also necessary when working at the module level. Path compression applies to `Texpand` too, i.e. only one `Texpand` is kept after calling `repr`.

There is a tension between two interpretation of types:
* for unification, expansion is completely transparent, and `ty` and `ty'` should actually be seen as identical. We call it the semantical view.
* for modules, the same signature can be given different meanings according to the environment, so that we need to revert to the `Tconstr` view before applying substitutions for instance. We call it the syntactical view.

In this PR the two views are not clearly separated, but `expand_abbrev` is given a new parameter `link`.
* if `link` is `true`, `expand_abbrev` sets `ty` to `Texpand(ty',path,args)`, and returns `ty` rather than `ty'`
* if `link` is `false`, `expand_abbrev` leaves `ty` untouched and returns `ty'`, as before

The choice of calling `expand_head` (`link = true`) or `expand_head_nolink` (`link = false`) depends on the context. In most cases, and in particularly during unification and in `Typecore`, one calls `expand_head`, while in the module system and a few required places one calls `expand_head_nolink`.

Once a `Texpand` is introduced by `expand_abbrev`,
* the expansion and abbreviation are kept by `Ctype.copy`
* but the expansions is discarded by `Subst.typexp`, i.e. `Texpand(ty,path,args)` is copied as `Tconstr(path,args,ref Mnil)`.

This is a rather big change, and I expect the review to take a while. It took me about 6 months to get it right.
One could even discuss whether it is worth bothering about it. My answer is that it is required if we want the current type checker to be correct. It is also interesting that it revealed some tensions between unification and modules.

Inspection of the testsuite reveals a variety of limitations:
* the validity check for abbreviations is done a posteriori, and currently we discard all abbreviations such that the expansion is itself included in one of the arguments; however in some cases this makes types harder to read
* the criterion for well-foundedness of type definitions was very syntactic and appears hard to enforce precisely in this semi-semantical approach, i.e. some types definitions that were rejected are now accepted

Fixes: #9314, #13230, #13881, #14036, #14307

garrigue commented:
Rebased on trunk without difficulties.
This doesn't change the issues with the printing of types in errors and the permissiveness with recursive type definitions, but at least the code seems reasonably stable.

nojb commented:
@garrigue: apologies for the delay (I had promised to test this branch on our codebase several months ago). Today I got around to doing it. I had to rebase it on the 5.1 branch, which was not completely trivial and I may have introduced some bugs along the way. Anyway, I was able to isolate the following program:
```
module type S = sig
  type db
  type db_storage
end

module Make(X: S) = struct
  type db = X.db
  type db_storage = X.db_storage
end

module M = Make (struct type db = unit type db_storage = db end)
```
which sends the compiler in this branch into an infinite loop.

garrigue commented:
@nojb Thank you for testing.
The same problem appears in all branches of this PR, so this is easily reproduced.
This is rather mysterious as this code does not seem to contain any recursion, I'm going to look into that.

garrigue commented:
@nojb Actually this one was easy to fix. A wrong call to the expanding version of `try_expand_safe` in `nondep_type_rec`.
The way it is easy to get things wrong is disturbing though.

nojb commented:
> @nojb Actually this one was easy to fix. A wrong call to the expanding version of `try_expand_safe` in `nondep_type_rec`. The way it is easy to get things wrong is disturbing though.

Thanks! I will try to do a bit more testing and report back.

gasche commented on typing/ctype.ml:
```diff
@@ -1694,18 +1731,22 @@ let rec try_expand_head
   | Escape e -> raise_for Unify (Escape e)
 
 (* Safe version of expand_head, never fails *)
-let expand_head env ty =
```
Could `link` (here and everywhere else) be a labelled parameter, to make callsites easier to read?

garrigue replied:
Done.

gasche commented on typing/parmatch.ml:
```diff
@@ -360,7 +360,7 @@ let clean_copy ty =
 
 let get_constructor_type_path ty tenv =
-  let ty = Ctype.expand_head tenv (clean_copy ty) in
+  let ty = Ctype.expand_head_nolink tenv (clean_copy ty) in
```
Why do we want to use the `_nolink` variant here?

garrigue replied:
Rule of thumb: the linking version is only needed to make unification stable.
Everywhere else, it is probably better to use the non-linking version, to avoid modifying the underlying type (however, constrained parameters rely on unification, so there is no guarantee that linking won't happen inside the type).

gasche commented on typing/printtyp.ml:
```diff
@@ -627,10 +735,17 @@ let rec uniq = function
+  match get_expand ty with
+    Some (path, args)
+    when not (is_Tvar ty || List.exists (deep_occur ty) args) ->
+      Tconstr (path, args, ref Mnil)
```
This code is the same as in Subst.typexp. Should the two versions be factored as a common function, for example `get_unexpanded_desc`?

gasche replied:
Note: the `is_Tvar ty` part is specific to this present callsite. I would guess that the idea is to not-unexpand (keep the expanded result) when the expansion is "small", but a comment to indicate that this is indeed the intent would help.

garrigue replied:
I did as suggested, but this turned out to be a little tricky, as there are distinct special cases.

gasche commented on typing/subst.ml:
```diff
@@ -222,7 +221,13 @@ let apply_type_function params args body =
+    match get_desc ty, get_expand ty with
+      Tsubst _ as desc, _ -> desc
+    | _, Some (path, args) when not (List.exists (deep_occur ty) args) ->
+        Tconstr (path, args, ref Mnil)
```
Can you explain why there is this guard against recursive (un)expansions? (This should probably be a comment in the code.)

garrigue replied:
The important part is the `Tsubst` case, which indicates that this node was already copied. There is an extra subtlety as linking to `Tsubst` does not erase the recorded abbreviation, so that one should look for `Tsubst` first. Will add a comment in the shared code.

gasche commented on typing/typedecl.ml:
```diff
@@ -783,15 +783,23 @@ let check_abbrev env sdecl (id, decl) =
+      | None, None -> true
+      | _ -> false
+    in
+    if TypeSet.exists check_parent parents then begin
```
This change is a bit surprising:
- performance-wise we replace a logarithmic check by a linear check
  (this could be optimised by performing a `TypeSet.find` first, and then checking the same-expand condition on the result)
- correctness-wise this is a bit dubious, as the `parents` set will store at most one element per repr-equivalence class, so distinguishing them is probably not going to work (do we preserve termination for example, if two types that are repr-equivalent but not check_parent-equivalent alternate with each other?)

smuenzel replied:
Regarding your second comment, is this what is happening in this example?
https://github.com/ocaml/ocaml/issues/9314#issuecomment-3112585011
```ocaml
type t = int s
and 'a s = 'a constraint 'a = t
```

smuenzel replied:
I think we will need to replace this check (proposal for a new version here: https://github.com/smuenzel/ocaml/pull/1)

gasche commented on typing/typedecl.ml:
```diff
@@ -814,6 +822,17 @@ let check_well_founded env loc path to_check visited ty0 =
+        let parents =
+          if rec_ok then TypeSet.empty else TypeSet.add ty parents in
+        List.iter (check_subtype parents trace ty) args)
+      ty;
```
I wonder what this piece of new code is doing.

gasche commented on typing/ctype.ml:
```diff
@@ -2510,6 +2533,7 @@ let find_lowest_level ty =
       iter_type_expr find ty
     end
   in find ty; unmark_type ty; !lowest
+*)
```
It would be nice to remove this function if it is not used anymore.

gasche commented on typing/ctype.ml:
```diff
@@ -885,26 +884,45 @@ let rec update_level env level expand ty =
         raise_escape_exn Self
     | _ ->
         set_level ty level;
         (* XXX what about abbreviations in Tconstr ? *)
```
@Octachron wonders if this comment is still accurate

garrigue replied:
It seems that I added this comment when adding the code for the relaxed value restriction.
It is still kind of mysterious to me: `expand_abbrev` calls `update_level` when using memoized abbreviations, but is it really sufficient?
Since this is orthogonal to the use of `Texpand`, the comment itself is still valid, but it seems that the code is correct anyway, so maybe we could remove the comment from trunk.

gasche reviewed (commented):
@Octachron and myself had a look at this PR today. We did not try to understand the change in depth, we looked at the diff in various places and wondered if we understand the change locally. So the result is a bunch of unrelated questions/comments on parts that we had trouble understanding.

I wonder if some of the changes here make the type-checker noticeably slower. (For example there are now extra occur checks, and the strategy to handle `Texpand` during link compression is naive.)

smuenzel commented:
@garrigue, have you had a chance to look at this PR again?

garrigue commented:
I need first to find time to rebase it, since it is sensitive to other changes.

garrigue committed:
Expand type in place, keeping original abbreviation

use non-linking expansion in includecore

accept ambivalence being strict even in non-principal mode

add slowdown example, make it a bit faster

fix interaction between Texpand and update_level

fix bug found by @nojb in nondep_type_rec

fix printing of constructor names in toplevel

fix interaction with modular explicits (losing some abbreviations)

Rename functions accessing abbrev_memo:
  iter_abbrev -> iter_abbrev_memo
  forget_abbrev -> forget_abbrev_memo
  cleanup_abbrev -> cleanup_abbrev_memo

simplify repr using repr_link / repr_expand

do not hide expansion of polymorphic variant types

smuenzel commented:
I've made an attempt at rebasing this onto trunk here: https://github.com/ocaml/ocaml/compare/trunk...smuenzel:keep-expansion-rebase3 

One interesting side-effect of this PR is that certain paths are no longer printed in the nice short form.
e.g. in testsuite/tests/lib-either/test.ml
```ocaml
 [left 1; right true];;
 [%%expect {|
- : (int, bool) Either.t list =
[Stdlib__Either.Left 1; Stdlib__Either.Right true]
 |}];;
```

I've ran all the tests in the rebased version, so you can take a look at the last commit for that diff.

smuenzel commented:
I attempted to fix the path printing issue here:
https://github.com/ocaml/ocaml/commit/c554cacdee32553d1b9d13e649273a57255af17a (I added this to my  rebase branch).
Basically, we can no longer rely on path comparison for printing constructors and labels, we must check the types themselves.

smuenzel committed:
Fix printing of types

Print Texpand if relevant

Add example from #14307

Fix printing of row types when they are printed as the result of an
expansion

Better printing of expanded types

smuenzel commented on typing/types.ml:
```diff
@@ -497,31 +501,64 @@ let commu_var () = Cvar {commu=Cunknown}
+      repr_link1 t d t'
+  | Tfield (_, k, _, t') when field_kind_internal_repr k = FKabsent ->
+      repr_link1 t d t'
+  | _ -> t
 
 (* scope_field and marks *)
```
Maybe we can introduce the following function to remove the usages of `ignore` below? It also documents the intention to normalize the `type_expr`

```suggestion
  | _ -> t

let normalize t =
  (ignore : type_expr -> unit) (repr t)

(* scope_field and marks *)
```

garrigue replied:
I do not think it is really needed in this file. It might be if we to do that outside of this file, but I am not sure there are such occurrences.

smuenzel committed:
Update testsuite

smuenzel committed:
new tests and fixes

Look up abbreviation in Ctype.extract_concrete_typedecl
Add record test for missing cmi

Use set_object_name with correct path

Fix PR5914 regression

Add test case for #13230

Add test for #13881

smuenzel commented on typing/btype.ml:
```diff
@@ -773,3 +782,44 @@ let instance_variable_type label sign =
+      (* Only re-instate an abbreviation if there is no risk to hide
+         something *)
+      match get_expand ty with
+      | Some (path, args) when not (List.exists (deep_occur ty) args) ->
```
Is `List.exists (deep_occur ty) args` different than `deep_occur_list`?

garrigue replied:
`deep_occur_list ty` is only supposed to be a slightly more efficient version of `List.exists (deep_occur ty)`, avoiding looking twice at the shared parts. Not using it here looks like an oversight (but better always test).

garrigue replied:
Done

smuenzel commented on typing/ctype.ml:
```diff
@@ -857,6 +852,14 @@ let update_scope_for tr_exn scope ty =
+let needs_expand env level path args =
+  let variance =
+    try (Env.find_type path env).type_variance
+    with Not_found -> List.map (fun _ -> Variance.unknown) args in
```
If we can't find the type, should we directly return `false`?

garrigue replied:
Good point. This could gain us a few msec.

smuenzel commented on typing/ctype.ml:
```diff
@@ -2754,9 +2770,15 @@ let rec unify uenv t1 t2 =
     type_changed := true;
     begin match (get_desc t1, get_desc t2) with
+    (* Using deep_occur here causes non-termination
+       in pr9314.ml and constraints.ml
```
Do we still need this once we have the well-founded changes? I believe these tests are no longer problematic then

garrigue replied:
Have to check. trunk uses `deep_occur` here, but is not able to handle these examples.

garrigue replied:
The non-termination is still there.

smuenzel replied:
For reference, it's this program that causes the `try_expand_head` loop
```ocaml
module type S =
sig
  type !'a s
  type !'a t = 'b  constraint 'a = 'b s
end
```

smuenzel commented on typing/out_type.ml:
```diff
@@ -488,10 +488,12 @@ let rec uniq = function
     [] -> true
   | a :: l -> not (List.memq (a : int) l) && uniq l
 
+let printer_get_desc ty = get_folded_desc ~keep_Tvar:true ty
```
Should we just say `let get_desc = ...`, since usage of `get_desc` in this file when we have `printer_get_desc` is probably an error.

garrigue replied:
Actually, there are some occurrences of `get_desc` in this file, and they make sense.
Also, too much overloading of identifiers might obscure things.
If we decide to prohibit the use of `get_desc` here, I would rather add the line
```ocaml
let get_desc = ()
```

smuenzel commented on typing/types.ml:
```diff
@@ -497,31 +501,64 @@ let commu_var () = Cvar {commu=Cunknown}
+   For instance, with "t1 -> t2" for [t1.desc = Tlink t2] and
+   "t1 -(path,args)-> t2" for [t1.desc = Texpand (t2, path, args)], we have
+   Before:
+      t -> t1 -(path1,args2)-> t2 -> t3 -(path2,args2)-> t'
```
```suggestion
      t -> t1 -(path1,args1)-> t2 -> t3 -(path2,args2)-> t'
```

garrigue replied:
Right

smuenzel commented on testsuite/tests/typing-misc/pr7937.ml:
```diff
@@ -2,25 +2,28 @@
 
+(* The examples here are broken by changes in unification,
+   because they force expansion *)
+
```
Is this just an issue of the error message being not helpful?
Or is there is a more fundamental problem?

garrigue replied:
The position of the error suggest something more fundamental is broken: the error used to be on the return value, but is now on the pattern. So if we replace the return value with `assert false` we may get a difference. Have to check.

smuenzel commented on typing/types.ml:
```diff
@@ -497,31 +501,64 @@ let commu_var () = Cvar {commu=Cunknown}
+   [Tlink] nor [Texpand]).
+   [repr t] returns the actual node [t'], and also normalizes [t]
+   so that it is either [t'] itself (if there was no prefix) or
+   either [Tlink t'] or [Texpand (t', path, args)] where [path] and [args]
```
Are there any cases where it's necessary to keep more than just the first `Texpand` in the chain? It might help to put something short in this comment here to explain why it's okay.

I particular, there are some functions that may remove an out of scope `Texpand`, such as `update_level_get_expand` -- is it possible that there is a `Texpand` later in the chain (if we didn't normalize) that would still be relevant?

garrigue replied:
Indeed, keeping only one `Texpand` allows more compression, and simplifies algorithms (`get_expand` returns only one value), but it is losing some information.
In some cases the `Texpand` we kept might be later invalidated, so that we keep no valid abbreviation for a type, and have to revert to the expanded form. For printing, this is only a mild nuisance. Whether it has worse consequences in the interaction with modules remains to be seen (I could not find a counter-example in that context).

smuenzel commented on typing/ctype.ml:
```diff
@@ -1802,6 +1842,7 @@ let full_expand ~may_forget_scope env ty =
         end
     else expand_head env ty
   in
+  let ty = Transient_expr.(type_expr (repr ty)) in (* forget Texpand *)
```
Why do we need forget here? We just called `expand_head_unif`/`expand_head` -- should these just be called with `~link:false`?

garrigue replied:
IIRC, this function is only used for printing. So experimenting and seeing whether it changes anything might be enough.

smuenzel commented on typing/ctype.ml:
```diff
@@ -2802,68 +2830,56 @@ and unify2_rec uenv t10 t1 t20 t2 =
+    unify3 uenv t1 t2
   else
-    try unify3 uenv t2 t2' t1 t1' with Unify_trace trace ->
+    try unify3 uenv t2 t1 with Unify_trace trace ->
```
Do we always want to call `unify3` with `Texpand` in the second position? If so, why not do the swap in `unify3` itself?

In addition, what is the reason for the asymmetry? `unify3` seems mostly symmetric in its arguments to me.

garrigue replied:
Interesting point. The logic here was more complicated before keep_expansion.
Need to compare with trunk, in which this asymmetry is really needed, and see whether the need has disappeared.

smuenzel commented on typing/ctype.ml:
```diff
@@ -2884,7 +2900,11 @@ and unify3 uenv t1 t1' t2 t2' =
             without_assume_injective uenv (fun uenv -> unify_list uenv tl1 tl2)
           else if in_current_module p1 (* || in_pervasives p1 *)
-               || List.exists (expands_to_datatype (get_env uenv)) [t1'; t1; t2]
+               || List.exists
+                   (fun (p, _) -> expands_to_datatype (get_env uenv) p)
+                   (Option.to_list (get_expand t1') @
+                    Option.to_list (get_expand t2') @
+                    [p1, []])
```
Maybe it's better for performance not to do the list concatenation, and just separately check each item?

garrigue replied:
Since `expands_to_datatype` calls `Env.find_type`, I believe this should not make a significant difference.

smuenzel commented:
@garrigue, @gasche has asked me to take a look at reviewing this implementation. I have some preliminary comments already, but it will take me additional time to comment on  the actual functionality, once I've understood it well enough.

Is this example (from your comment in #9314) a good one to step through the unification changes vs trunk?
```ocaml
type 'a gamma = 'b constraint 'a = < gamma : 'b >
let o : < gamma : 'a gamma > as 'a = object method gamma = 1 end
```

smuenzel commented on typing/types.mli:
```diff
@@ -241,6 +245,14 @@ val try_mark_node: type_mark -> type_expr -> bool
+  (* If [get_expand ty = Some (_, path, _)] then return the scope of [path]
+     otherwise [Ident.lowest_scope] *)
+val iter_expand: (Path.t -> type_expr list -> unit) -> type_expr -> unit
+val forget_expand: type_expr -> unit
```
This is the naming problem you mentioned (we're not getting the expansion, but the abbreviation).
Maybe something like `get_kept_abbrev` will do for a name? (based on the comment above, and the PR title)

garrigue replied:
Still thinking about it.
I would really like to call it `get_abbrev` (and call the constructor `Tabbrev`) but I think the name is already used. Maybe the original one can be renamed :-)

smuenzel commented:
> One could even discuss whether it is worth bothering about it. My answer is that it is required if we want the current type checker to be correct.

Regarding the importance of this feature -- I agree that this is important. There are a number of past bug reports that we can squash with this PR. I will attempt to make a list.

I have to thank you for undertaking this very involved work -- it seems to be incredibly tricky to get right.

garrigue commented:
Thank you for your review. But it might be a good idea to rather review your rebased code.
However we have to do a cross-review: you review my code and I review your code.
Not sure how to proceed.

As for your question about unification, #9314 is indeed one motivation, but the idea of the code is just that unification should share the expanded forms, rather than share the abbreviated form while keeping the expanded forms partly unshared. The abbreviations are only kept for printing types, and for compatibility at the module level. This question is clearer in examples about ambivalence, for instance #14307.

smuenzel commented:
> But it might be a good idea to rather review your rebased code.
> However we have to do a cross-review: you review my code and I review your code.
> Not sure how to proceed.

Maybe you can pull my rebase (without the well-founded changes) into this PR, and then we can cross-review.
Will we need to find another reviewer as well (since we're now co-authors?)? I'm not exactly sure what the review policy is.

gasche commented:
The general policy is that someone with the "green approval bit" (here: a maintainer) that is not an author of the PR must approve it before it can get merged. Usually this means that this person has reviewed the PR in details themselves, but not always.

In the context of the current PR, we have failed to review it in reasonable amounts of time, because no one has the expertise to do so and the usual sacrificial lambs (@Octachron and myself) were busy with other things. We are very lucky that you feel motivated to review it, and looking for a second competent review would put us back in the same unfortunate situation as before. My current assessment is that if we get to a state where you and Jacques both agree that the code is good enough to be merged, I would be happy to approve on your behalf without reviewing -- or probably with high-level discussions for example during a type-system meeting. That can of course change, in particular if you yourself end up concluding that there is too much of your own work in the PR for you to give a fair assessment of it.

garrigue commented:
Pushed the version rebased by @smuenzel .
Changes in  [keep-expansion-rebase3-well-founded](https://github.com/smuenzel/ocaml/tree/keep-expansion-rebase3-well-founded) not included yet.

garrigue commented on testsuite/tests/typing-modular-explicits/general.ml:
```diff
@@ -249,7 +249,7 @@ This argument cannot be applied with label "~c"
 
 [%%expect{|
-val x_from_struct : int = 3
+val x_from_struct : T.t = <external>
```
This looks like a bug: `T` does not exist in the current context.

garrigue replied:
I could somehow fix the interaction with modular explicits, but the way we forget abbreviations means some types are more expanded than before. This might become a problem when there are lots of modules.

garrigue commented on testsuite/tests/typing-modular-explicits/general.ml:
```diff
@@ -158,7 +158,7 @@ let apply_labelled_success = labelled' ~y:3
 val labelled' : (module M : Typ with type t = int) -> y:M.t -> M.t = <fun>
-val apply_labelled_success : (module Typ with type t = int) -> int = <fun>
+val apply_labelled_success : (module Typ with type t = int) -> M.t = <fun>
 |}]
```
Same bug as below. There is a problem with modular explicits.

smuenzel commented on typing/ctype.ml:
```diff
@@ -985,25 +983,44 @@ let rec update_level env level expand ty =
         let env = Env.add_module (Ident.of_unscoped id) Mp_present mty env in
         set_level ();
         update_level env level expand t
+          (* CR smuenzel : Anything to do for expand? *)
```
@garrigue I had left this note here for myself but have not yet investigated

smuenzel replied:
I believe this can be removed.

smuenzel commented on typing/ctype.ml:
```diff
@@ -3231,18 +3252,24 @@ let rec unify uenv t1 t2 =
         update_scope_for Unify (get_scope t1) t2;
         link_type t1 t2
-    | (Tconstr (p1, [], a1), Tconstr (p2, [], a2))
+    | (d1, d2) -> match (d1, d2, get_constr_desc t1, get_constr_desc t2) with
```
Is my assumption right that this is the key line that introduces the `Texpand` information into unification (thus allowing us to revert to the unexpanded view when we could use it for constructor unification)?

garrigue replied:
Not really. This line handles the special case of unification between type constructors that do not have parameters. This is particular covers locally abstract types (`type_is_newtype = true`), and is important for GADTs. Namely, in that case we do not want to force complete expansion, as it might cause ambivalence, but rather do expansion incrementally, looking for a common type where we can stop expanding.

smuenzel commented on typing/types.ml:
```diff
@@ -536,31 +540,64 @@ let commu_var () = Cvar {commu=Cunknown}
+  | Tfield (_, k, _, t') when field_kind_internal_repr k = FKabsent ->
+      repr_link1 t d t'
+  | _ -> t
 
```
Maybe something like this to remove the allocation:
```ocaml
let repr_update ~t_orig d =
  log_change (Ccompress (t_orig, t_orig.desc, d));
  t_orig.desc <- d

let rec repr_expand ~depth ~path ~args ~t_orig t' =
  match t'.desc with
  | Tlink t' | Texpand (t', _, _) ->
      repr_expand ~depth:(depth + 1) ~path ~args ~t_orig t'
  | _ when depth > 1 ->
      repr_update ~t_orig (Texpand (t', path, args));
      t'
  | _ -> t'

let rec repr_link ~depth ~t_orig d t' =
  match t'.desc with
  | Tlink t' as d ->
      repr_link ~depth:(depth + 1) ~t_orig d t'
  | Texpand (t', path, args) ->
      repr_expand ~depth:(depth + 1) ~path ~args ~t_orig t'
  | Tfield (_, k, _, t') as d
    when field_kind_internal_repr k = FKabsent ->
      repr_link ~depth:(depth + 1) ~t_orig d t'
  | _ when depth > 1 ->
      repr_update ~t_orig d;
      t'
  | _ -> t'

let repr t =
  repr_link ~depth:0 ~t_orig:t t.desc t

```

garrigue replied:
Indeed, we could do something like that.
I was waiting for the code to stabilize, and to do some actual benchmarks first.

garrigue replied:
I followed this idea, as it looks cleaner anyway.
But there does not seem to be a clear improvement in performance.

smuenzel replied:
I also experimented with using a deeper match in the top-level `repr` function, to avoid calls to `repr_link`/`repr_expand` when there is nothing to do, but didn't see a performance difference

smuenzel commented:
Here's the fix for printing the row type: https://github.com/COCTI/ocaml/compare/keep-expansion...smuenzel:ocaml:keep-expansion  (https://github.com/COCTI/ocaml/commit/4e4efb6e8ea492ccdf3b7496c88c587e27d24cf4)
We were dropping the constructor name when unaliasing the expanded type when printing.

smuenzel commented on typing/out_type.ml:
```diff
@@ -2032,41 +2033,77 @@ let same_path t t' =
-  | _ ->
-      false
+  | Tconstr _, _ | _, Tconstr _ -> false
+  | _ -> true
```
I believe this should be `false` to fix `testsuite/tests/typing-gadts/pr5689.ml`, however, `testsuite/tests/typing-objects/Exemples.ml` ends up looking strange with this change, printing types like:
```ocaml
         "#comparable as 'a" = "< cmp : 'a -> int; .. >"
```
Most likely `'a` and `'a` are distinct here, but some other issue in printing is assigning the same name, possibly to do with preparation

smuenzel replied:
Actually, it seems that this is indeed the expected type, and it has nothing to do with any issues in printing.

So these changes should fix `pr5689.ml`: https://github.com/smuenzel/ocaml/commit/30111fad7a0f6becc09bac36165d0f17697a56c0

garrigue replied:
I merged these changes in my `keep_expansion` branch.
Note that I removed the unneeded line in `Out_type.same_path`.

smuenzel commented on typing/includemod.ml:
```diff
@@ -242,6 +242,7 @@ module Core_inclusion = struct
   let value_descriptions ~loc env ~direction subst id vd1 vd2 =
     if Directionality.mark_as_used direction then
       Env.mark_value_used vd1.val_uid;
+    let vd1 = Subst.value_description Subst.identity vd1 in
```
In typedecl, there is a comment saying
` (* Copy the type declarations to remove spurious expansions *)`
(and also another ` (* Using [Subst] reverts expansions *)`)
when we're using the identity substitution. I assume this is the intention here as well?

garrigue replied:
Indeed.

smuenzel commented on typing/btype.ml:
```diff
@@ -823,3 +831,45 @@ let instance_variable_type label sign =
+  if get_level ty >= get_level t0 && try_mark_node mark ty then begin
+    if eq_type ty t0 then raise Occur;
+    iter_type_expr (deep_occur_rec mark t0) ty;
+    iter_abbrev (fun _p tyl -> List.iter (deep_occur_rec mark t0) tyl) ty
```
Do we need to look at the level (scope) of `_p`? Or is it not necessary due to some invariant?

garrigue replied:
This one just checks for occurrences, so the scope is not directly relevant.
And anyway, we do not have a notion of current level here, so we have nothing to compare to.
Same problem in `get_folded_desc` below.

smuenzel commented:
I'm adding some more questions as I review, it's taking me some time to understand what's going on.

FWIW, I'm not that familiar with the code here, so please excuse the large number of (sometimes basic) questions.

smuenzel commented on typing/btype.ml:
```diff
@@ -823,3 +831,45 @@ let instance_variable_type label sign =
+
+let rec deep_occur_rec mark t0 ty =
+  if get_level ty >= get_level t0 && try_mark_node mark ty then begin
+    if eq_type ty t0 then raise Occur;
```
Regarding `eq_type`, we check `t` and `repr t`  against each other. But we don't check `Texpand` against `Tconstr`. Is that intentional? And if so, should we add a clarifying comment about the notion of equality that is being checked in `types.mli` ?

smuenzel replied:
PS: It's a bit confusing that there is `Types.eq_type` and `Ctype.eqtype`

smuenzel replied:
So my conclusion here is that:
- `eq_type` checks for type expression equality at the physical level (without examining the meaning of the type expression), *except* for normalizing the type expression and checking the `repr` as well. This gives a nice bounded-in-time check. Adding anything about `Texpand` here would introduce recursion that may be unbounded.
- `eqtype` recurses through the type description, so it is a kind of semantic comparison between type expressions (but it does not examine abbreviations)

garrigue replied:
The point is that with keep-expansion, we need only to compare the results of `repr`. For this, the abbreviation is irrelevant. The comparison of the types before calling `repr` is just an optimization.

garrigue reviewed (commented):
Just answering questions.
It is certainly fine to ask, the more you understand, the better the review.

smuenzel committed:
Update comments

smuenzel committed:
New well-founded implementation based on the definition

smuenzel committed:
Bootstrap

smuenzel commented on typing/ctype.ml:
```diff
@@ -3260,68 +3281,56 @@ and unify2_rec uenv t10 t1 t20 t2 =
-  let scope = Int.max (get_scope t1') (get_scope t2') in
+  ignore (expand_head_unif env t1);
+  ignore (expand_head_unif env t2);
+  ignore (expand_head_unif env t1);
+  ignore (expand_head_unif env t2);
```
What is the reason for this double expansion? If we only do one expansion, shouldn't this work as well (since if the second expansion is necessary, we would do so later)

The testsuite seems to be fine if we remove the second expansion -- are there any efficiency reasons?

smuenzel replied:
I traced it back to this: https://github.com/ocaml/ocaml/commit/8ff158b81ce34f534c2c94bcd07b3e842930d1ed (#6954)

But it seems like the code example there is also fine without the second expansion. (and it's also in the testsuite)
```ocaml
module X = struct module Y = struct module type S = sig type t end end end

(* open X (* works! *) *)
module Y = X.Y

module type Arg = sig
  module A : sig type 'a t constraint 'a = (module Y.S with type t = 'at) end
end

module MyA = struct
  (* type 'a t = 'at constraint 'a = (module X.Y.S with type t = 'at) (* works! *) *)
  type 'a t = 'at constraint 'a = (module Y.S with type t = 'at)
end
module Make (Arg : Arg with module A = MyA) = struct
  type t = (module X.Y.S with type t = unit)
  let x : t Arg.A.t = Obj.magic 0
end
```

I also checked another historical PR related to these lines, and it seems to be fine as well:
https://github.com/ocaml/ocaml/issues/8359


For what it's worth, after archeology, we get back to this change: https://github.com/ocaml/ocaml/commit/723a2e6c7c4f43baecdc5f8106208e62237fdf79, unfortunately, the commit message only says "Bug dans l'unification", without giving an example.

garrigue replied:
This double expansion is actually much older than that. It was already there in OCaml 2, more than 25 years ago. @vouillon added it.
There used to be a comment:
```ocaml
  (* Expansion may have changed the representative of the types... *)
```
Note that this may be actually unneeded now that one has to use `get_desc` to access the contents of a type, but to be sure one would have to check whether removing it in old version was breaking something in the testsuite at the time. Looking for something in OCaml 3.12 might be good enough.

vouillon replied:
I guess expanding a type can perform some unification. So, if we write just the following, `t1'` might no longer be expanded after the computation of `t2'`. But, because of caching, if we do the expansion a second time, we get the right result.
```ocaml
  let t1' = expand_head_unif env t1 in
  let t2' = expand_head_unif env t2 in
```

garrigue replied:
Actually we still have an error in trunk if we remove this double expansion, but keep-expansion does not seem to need it anymore. Would need a deeper analysis.

smuenzel commented:
> As for your question about unification, #9314 is indeed one motivation, but the idea of the code is just that unification should share the expanded forms, rather than share the abbreviated form while keeping the expanded forms partly unshared. The abbreviations are only kept for printing types, and for compatibility at the module level. This question is clearer in examples about ambivalence, for instance #14307.

Thanks for the pointer -- I think I now understand how this (#14307) works. In both trunk and keep-expansion, we initially are able to update the scope of `a` (which results from an expansion of `c` to `d -> a`) from `scope=0` to `scope=12`. But keep-expansion is able to share this scope update so later, when we're looking at `a` again, we will have remembered that `scope=12`, whereas trunk uses a different instance of the type (which can be seen from the different type id) with `scope=0`, so we are unable to detect the scope escape in trunk. 

(This is very apparent when looking at a diff of the calls to the unification functions side-by-side. It helps to remove the `deep_occur` condition from `unify` in trunk, so that the traces match up more close.)




Edit: for those interested in the details, seeing this helped me a lot (hope it's correct!)

Initially, we update the scope:
```
-> unify2_rec
  keep-expansion:
    on entry:
         t1={id=2249;level=15;scope=12;marks=0;desc=
             Tconstr(a/283[4],[],[b/284[6]])}
         t2={id=2217;level=15;scope=0;marks=0;desc=Tconstr(a/283[4],[],[])}
    on exit:
         t1=t2={id=2217;level=15;scope=12;marks=0;desc=
                Tconstr(a/283[4],[],[])}
  trunk:
    on entry:
         t1={id=2282;level=15;scope=12;marks=0;desc=
             Tconstr(a/283[4],[],[b/284[6]])}
         t1'={id=2266;level=15;scope=12;marks=0;desc=
              Tconstr(b/284[6],[],[b/284[6]])}
         t2=t2'={id=2278;level=15;scope=0;marks=0;desc=
                 Tconstr(a/283[4],[],[c/285[8]])}
    on exit:
         t1={id=2278;level=15;scope=12;marks=0;desc=
             Tconstr(a/283[4],[],[c/285[8]])}
         t1'={id=2266;level=15;scope=12;marks=0;desc=
              Tconstr(b/284[6],[],[b/284[6]])}
         t2=t2'={id=2278;level=15;scope=12;marks=0;desc=
                 Tconstr(a/283[4],[],[c/285[8]])}
```

But then later, we need to detect the scope escape (note `scope=12` vs `scope=0`):
```
-> unify1_var
  keep-expansion:
         t1={id=2111;level=11;scope=0;marks=0;desc=Tvar None}
         t2={id=2217;level=15;scope=12;marks=0;desc=Tconstr(a/283[4],[],[])}
  trunk:
         t1={id=2109;level=11;scope=0;marks=0;desc=Tvar None}
         t2={id=2220;level=15;scope=0;marks=0;desc=Tconstr(a/283[4],[],[])}
```

Note that we see type 2217 in the expansion of `b` and `c` in between the scope update and the scope escape detection:
```
         t1=t2={id=2243;level=15;scope=12;marks=0;desc=
                Texpand(
                 {id=2217;level=15;scope=12;marks=0;desc=
                  Tconstr(a/283[4],[],[])},
                 b/284[6],[])}
         t1=t2={id=2242;level=15;scope=11;marks=0;desc=
                Texpand(
                 {id=2216;level=15;scope=11;marks=0;desc=
                  Tarrow("",
                   {id=2218;level=15;scope=0;marks=0;desc=
                    Tpoly(
                     {id=2219;level=15;scope=0;marks=0;desc=
                      Tconstr(d/286[10],[],[])},[])},
                   {id=2217;level=15;scope=12;marks=0;desc=
                    Tconstr(a/283[4],[],[])},Cok)},
                 c/285[8],[])}
```

but in trunk we track the expansions separately, 
```
         t1=t1'=t2={id=2212;level=15;scope=11;marks=0;desc=
                    Tconstr(c/285[8],[],[c/285[8]])}
         t2'={id=2277;level=15;scope=11;marks=0;desc=
              Tarrow("",
               {id=2279;level=15;scope=0;marks=0;desc=
                Tpoly(
                 {id=2280;level=15;scope=0;marks=0;desc=
                  Tconstr(d/286[10],[],[c/285[8]])},[])},
               {id=2278;level=15;scope=12;marks=0;desc=
                Tconstr(a/283[4],[],[c/285[8]])},Cok)}
```
and just a couple of functions later (after the return from `unify2`???), we've lost the scope of `a`

smuenzel commented on typing/ctype.ml:
```diff
@@ -3260,68 +3281,56 @@ and unify2_rec uenv t10 t1 t20 t2 =
-  ignore (expand_head_unif env t2');
-  let t1' = expand_head_unif env t1' in
-  let t2' = expand_head_unif env t2' in
-  let lv = Int.min (get_level t1') (get_level t2') in
-  let scope = Int.max (get_scope t1') (get_scope t2') in
+  ignore (expand_head_unif env t1);
+  ignore (expand_head_unif env t2);
+  ignore (expand_head_unif env t1);
+  ignore (expand_head_unif env t2);
+  let lv = Int.min (get_level t1) (get_level t2) in
+  let scope = Int.max (get_scope t1) (get_scope t2) in
   update_level_for Unify env lv t2;
   update_level_for Unify env lv t1;
   update_scope_for Unify scope t2;
   update_scope_for Unify scope t1;
-  if unify_eq uenv t1' t2' then () else
-
-  let t1, t2 =
-    if !Clflags.principal
-    && (find_lowest_level t1' < lv || find_lowest_level t2' < lv) then
-      (* Expand abbreviations hiding a lower level *)
-      (* Should also do it for parameterized types, after unification... *)
-      (match get_desc t1 with Tconstr (_, [], _) -> t1' | _ -> t1),
-      (match get_desc t2 with Tconstr (_, [], _) -> t2' | _ -> t2)
-    else (t1, t2)
-  in
-  if unify_eq uenv t1 t1' || not (unify_eq uenv t2 t2') then
-    unify3 uenv t1 t1' t2 t2'
+
+  if unify_eq uenv t1 t2 then () else
+  if get_abbrev t1 = None || get_abbrev t2 <> None then
+    unify3 uenv t1 t2
   else
-    try unify3 uenv t2 t2' t1 t1' with Unify_trace trace ->
+    try unify3 uenv t2 t1 with Unify_trace trace ->
       raise_trace_for Unify (swap_trace trace)
 
-and unify3 uenv t1 t1' t2 t2' =
+and unify3 uenv t1' t2' =
```
Regarding the prime on the variables: Is the convention here that `t1'` or `t2'` should contain `Texpand`? Or just that we previously called `expand_head_unif`? 
(previously, we passed expanded and unexpanded forms separately)

garrigue replied:
This is not directly related to the `expand_head_unif`, but indeed it ensure that, if there was an expandable abbreviation, then it will appear as a `Texpand` in the type.
The goal of this code is to keep an abbreviation when one of `t1` or `t2` has one.
So we have this convention that, if either `t1` or `t2` is a `Texpand`, then `t2'` in `unify3` should be a `Texpand`. This said, this is not completely optimal, as we do not check whether this abbreviation is valid or not. But this appears sufficient to keep an abbreviation in most cases.

smuenzel commented on typing/types.ml:
```diff
@@ -800,30 +845,54 @@ let last_snapshot = Local_store.s_ref 0
+      Transient_expr.set_desc ty d
+  | _ ->
+      Transient_expr.set_desc ty (Tlink ty')
+  end;
```
Would it be wrong to `set_desc ty (Tlink ty')` and then call `repr`? I'm asking because we have a kind of semi-normalization here, and maybe just always calling `repr` would make things easier.

garrigue replied:
I suppose it would do the same thing, just be a bit less efficient (need to update twice, and note that we also log path-compression).

smuenzel commented on typing/types.ml:
```diff
@@ -46,6 +46,9 @@ and type_desc =
   | Tfunctor of arg_label * Ident.Unscoped.t * package * type_expr
+  | Texpand of type_expr * Path.t * type_expr list
+        (* NB: let's move Tlink and Tsubst (temporary nodes) to the end of
+           this definition *)
```
I assume in a separate PR?

garrigue replied:
I actually, I planned to wait for when a bootstrap is required, but this is not the case for this PR, so it might be better indeed to do it in a separate PR.

smuenzel replied:
The well-founded changes seem to require a bootstrap (or at least I wasn't able to get it to work without one). I could incorporate it there

smuenzel committed:
Expand types for printing errors in typecore

Much simpler solution for type printing

Remove stale comment

Better error message for well-founded parameters

smuenzel commented on testsuite/tests/typing-misc/occur_check.ml:
```diff
@@ -62,15 +62,26 @@ let wrong_to_seq (xt : 'a t) : 'a Seq.t =
 [%%expect{|
 type 'a t = T of 'a
-val wrong_to_seq : ('a Seq.t as 'a) Seq.t t -> 'a Seq.t Seq.t = <fun>
+Line 4, characters 2-22:
+4 |   Seq.cons Seq.empty x
+      ^^^^^^^^^^^^^^^^^^^^
+Error: This expression has type "'a Seq.t Seq.t" = "unit -> 'a Seq.t Seq.node"
+       but an expression was expected of type
+         "'a Seq.t Seq.t Seq.t" = "unit -> 'a Seq.t Seq.t Seq.node"
+       Type "'a Seq.t" = "unit -> 'a Seq.node" is not compatible with type
+         "'a Seq.t Seq.t" = "unit -> 'a Seq.t Seq.node"
+Hint: This function application is partial, maybe some arguments are missing.
```
I think we need to update the comment above, since this is now fixed.

(for reference: https://github.com/ocaml/ocaml/pull/12974/changes)

garrigue replied:
Done.

smuenzel committed:
Move Tlink and Tsubst to the end and make Mcons an inline record

smuenzel reviewed (approved):
I think I'm satisfied that I have a reasonable understanding and that this does what's advertised.

Regarding correctness, we have the following two possible classes of bugs:
1. We don't link an expansion when we need to: this leads to failures like in `ambivalent_apply.ml`, where we don't propagate levels/scope.
2. We link an expansion when we should not, and thus an expansion ends up somewhere where it's not valid. We don't have any examples of that (I think?), but we erase any expansions when operating in the module system (see `includemod`).

There are no obvious instances of either mistake that I could find, and all the testsuite results make sense.


What should our next steps be?
Some things we need to do:
1. There are some comments in tests that are no longer up to date (i.e. saying a test fails when it does not), so those need to be updated
2. If you are happy with the well-founded changes, you can probably pull them into this PR. (note that there's a bootstrap in there)
3. We should rebase the whole PR (although I don't think anything made it into trunk that could affect us)
4. Any more testing we should do (with code outside of the compiler)?
5. We should come up with a list of issues that are fixed. This is what I remember: 
	- #13230
	- #13881 
	- #9314
	- #14307
	- #14036
	- the issue mentioned in the comment in #12974

garrigue commented:
I merged https://github.com/smuenzel/ocaml/pull/1 .
Will now rebase.

garrigue committed:
update comments in tests and add Changes

test outputs for lib-either and match-side-effects

garrigue commented:
I have rebased and updated the comments in the tests.
It is basically ready for merging.
If you are purist there is still a question about the error messages for cyclic abbreviations, which only say "There is a cycle in ..." whereas this should probably be "This abbreviation is cyclic"

I am sure your list of related bugs is exhaustive, but it is hard to find them.
They might appear progressively.
I should at least update the PR description, to include your name and the list of fixed bugs, since this is usually what gets in the merge log.

Also, I think the idea was that you could ask question if you think some explanation is needed in the code. This can be done now or later, but I understand that it may be cryptic.

garrigue commented:
This PR is now ready for merging, however more tests could help.

@Octachron Can we test it on opam? I still do not know how to do that :-(

@nojb Could you again try it on Lexifi's code base?
We have added extra tests to the testsuite, but it would still be good to test it on a large codebase.
Also, the behavior in case of error matters too, to see whether error messages were degraded.

@goldfirere Do you know if anybody could test this at Jane Street?
Again, the behavior with lots modules of functors has not been extensively tested yet.

nojb commented:
> @nojb Could you again try it on Lexifi's code base?

In order to do this, I will need to have a version of the patch rebased on 5.4.1. I will give it a try, but I may end up needing help with the rebase.

garrigue commented:
> > @nojb Could you again try it on Lexifi's code base?
> 
> In order to do this, I will need to have a version of the patch rebased on 5.4.1. I will give it a try, but I may end up needing help with the rebase.

I will do it now. There are probably going to be a few conflicts.

nojb commented:
> > > @nojb Could you again try it on Lexifi's code base?
> > 
> > 
> > In order to do this, I will need to have a version of the patch rebased on 5.4.1. I will give it a try, but I may end up needing help with the rebase.
> 
> I will do it now. There are probably going to be a few conflicts.

Thanks! That's appreciated.

garrigue commented:
Rebase of 5.4 done. Was a bit harder than expected, but tests seem fine.

smuenzel commented on Changes:
```diff
@@ -76,6 +90,12 @@ Working version
+* #?????: Reimplement wellfoundedness checks based on the definition of
+  wellfoundedness in the typechecker comments.
+  (Stefan Muenzel, review by ??????)
+
```
```suggestion
```

smuenzel commented on typing/typedecl.ml:
```diff
@@ -925,27 +1050,29 @@ let check_well_founded_manifest ~abs_env env loc path decl =
    (we don't have an example at hand where it is necessary), but we
    are doing it anyway out of caution.
```
```suggestion
```

smuenzel replied:
That didn't really come out right in the github suggestion. I suggest removing the part of the comment that's no longer true

smuenzel commented:
> I have rebased and updated the comments in the tests. 

Thank you!

>It is basically ready for merging. If you are purist there is still a question about the error messages for cyclic abbreviations, which only say "There is a cycle in ..." whereas this should probably be "This abbreviation is cyclic"

I have no strong opinion on the phrasing of the message. The users should be able to tell what is wrong with either one.

> I am sure your list of related bugs is exhaustive, but it is hard to find them. They might appear progressively. I should at least update the PR description, to include your name and the list of fixed bugs, since this is usually what gets in the merge log.

I see you've done this, thanks! There's still an old entry from me in the changes file, I added a github suggestion to remove it.

> 
> Also, I think the idea was that you could ask question if you think some explanation is needed in the code. This can be done now or later, but I understand that it may be cryptic.

I read through the code again, and the only comments I thought to add you have already incorporated when you pulled in my latest changes. I think I understand now why there aren't that many comments in general: in many cases, there isn't much to say about a function or type locally, but the more difficult part is in how things tie together. It's possible there are some good spots to add deeper comments (such as the one about well-foundedness that you and @gasche worked on -- thanks for that, couldn't have done my part without it!), and I will think about that and maybe create a separate PR.

smuenzel commented:
In case you would like to change the phrasing to "is cyclic":
https://github.com/COCTI/ocaml/compare/keep-expansion...smuenzel:ocaml:keep-expansion-cyclic-error-message

gasche commented:
My understanding is that #14673 (the 5.4 rebase) is revealing some leftover bugs in ecosystem-wide testing. I suppose that one should wait for those to be addressed before merging here as well. (Unless you believe that this comes from the 5.4 rebase only.)

smuenzel commented:
> My understanding is that #14673 (the 5.4 rebase) is revealing some leftover bugs in ecosystem-wide testing. I suppose that one should wait for those to be addressed before merging here as well. (Unless you believe that this comes from the 5.4 rebase only.)

Yes, I think it's correct to resolve them first.

garrigue committed:
Various fixes for keep-expansion compiling opam

* Fix update_level using check_level_type_rec

* In Subst.typexp, do not link to folded desc

* Fix expansion and use get_folded_desc in nondep_type_rec

* Make output type parsable for named closed polymorphic variant

They fix respectively base_quickcheck, lablgtk, core_kernel,
and js_of_ocaml (cf #14573)

test output

smuenzel commented on typing/ctype.ml:
If we have `Texpand` at the head of `ty`, isn't it possible that this expansion would fail, because the type is already expanded?

garrigue replied:
Indeed, if `desc != get_desc ty`, then one should just remove the abbreviation with `forget_abbrev` and call recursively. Need also to remove the node from `nondep_hash` before recursing.

garrigue reviewed (commented):
Still trying.

garrigue commented:
I thought that now everything was fine, but after cherry-picking changes from keep-expansion-5.4 (#14673) and rebasing, a new bug appears in relation with mutually recursive classes.
(See cocti#19 for the rebased branch)
It probably comes from an interaction with #14680, but only appears in the toplevel or expect tests :-)

nojb commented:
I have an unrelated question concerning this development: in the fork of OCaml that we maintain inside LexiFi we extended the type algebra to keep track of attributes at the type level, eg `type t = int [@t foo="bar"]` (where the attribute `[@t foo="bar"]` "flows" along the type `t`). We implemented this using some horrible hacks to avoid actually changing the representation of types (`Types.type_expr`). Now that we have `Texpand _`, I wonder if it could make sense to also keep track of any attributes found when expanding a manifest type in an extra field of this node. This would be unused by the compiler (today), but it could be useful later (and it looks like it would simplify the implementation of our fork). Not suggesting to do this right now, of course, but I would be interested to hear if you have any opinions in this regard?

garrigue commented:
@nojb I suppose it would make sense, in that `Texpand` nodes are supposed to be kept around as long as possible.
Not however that this is still "best effort". We only keep on abbreviation, and discard it when we are out of scope, and when there is a recursion between parameters and the whole type.

smuenzel commented:
> but only appears in the toplevel or expect tests :-)

Indeed, it's a little weird. The following is okay:

```ocaml
type 'a x = 'a lst
let q (x : _ x) = x#print Format.print_int;;
q l1;;      
(3::10::[])- : unit = ()
```
But this is not:
```ocaml
l1#print Format.print_int;;
Error: This expression is not an object; it has type int lst
```

Also, this:

```ocaml
(l1 : int x)#print Format.print_int;;
Error: The value l1 has type int lst but an expression was expected of type
         int x
       Type lst is abstract because no corresponding cmi file was found
       in path.
       Type lst is abstract because no corresponding cmi file was found
       in path.
```

FWIW, the type looks like:
```
> +Error: This expression is not an object;
> +       it has type "int lst" ({id=25073;level=15;scope=0;marks=0;desc=
> +                             Texpand(
> +                              {id=25075;level=15;scope=0;marks=0;desc=
> +                               Tconstr(lst/1050[12],
> +                                [{id=25074;level=15;scope=0;marks=0;desc=
> +                                  Tconstr(int/1!,[],[])}],[])},
> +                              lst/1048[12],[{id=25074}])})

```

smuenzel commented:
I believe we are somehow modifying the type desc of `lst` when we print the signature of the class. 

`Out_type.Variable_names.reserve` calls `Ctype.normalize_type_rec`, which can modify the type desc from a `Tobject` to a `Tconstr`.

No clue how to fix this yet. 

FWIW: deleting the call to `normalize_type` in `Out_type.Variable_names.reserve` makes the test pass.

But I'm very suspicious that the changes relating to `set_object_name` in #14680 are the core of the issue

smuenzel commented:
See here for the fix:

https://github.com/COCTI/ocaml/compare/keep-expansion-20260406...smuenzel:ocaml:keep-expansion-20260406-set-object-name

The issue was that the call to `set_object_name` was using the wrong path.

FYI @Octachron

garrigue commented:
Just for the record, the modification added for the 5.4 branch seem to have no impact on performance: typechecking times are still on par with trunk.

garrigue commented:
> See here for the fix:

Great job @smuenzel !
I must admit that the inability to use my preferred tool, the debugger, was seriously impairing my motivation.
The change of behavior according to whether we print types or not was indeed a strong clue that something wrong was happening in `normalize_type`, however tracing it back to a wrong path in `typeclass` is even more impressive :-)

I included your fix, and we are now ready for merging.

This said, while `normalize_type` helped discover a bug this time, it is a dirty hack that I introduced long ago for polymorphic variants, and should probably go away. This is a cause of instability.

smuenzel commented:
Sounds good!

Do you also want to remove this comment before merge: https://github.com/COCTI/ocaml/compare/keep-expansion...smuenzel:ocaml:keep-expansion-final ?

PS: I also added a small improvement to the printing of well-founded error messages with parameters to that branch

garrigue commented:
I have included your 3 last commits, and reordered/squashed the commits so that we can merge keeping a reasonable number of separate commits.
You can have a last look before I merge.

garrigue commented on testsuite/tests/prim-alias/recursive_modules.ml:
```diff
@@ -89,5 +89,5 @@ Line 5, characters 19-25:
 Error: The type of this alias does not match that of the aliased primitive.
        Type "a -> b" is not compatible with type "'a -> 'a"
-       Type "b" is not compatible with type "'a"
+       Type "b" is not compatible with type "'a" = "a"
```
This expansion is a bit strange. How can a type variable be expanded into a path?
Or is the meaning different.

smuenzel replied:
I'm not sure what is happening here, I assume something to do with printing. But I do believe that we keep the order of types correctly from left to right. I don't know if it's important enough to figure out or if we should look at improving printing in a follow up (since, as far as I can tell, it doesn't affect correctness, and there is sufficient information in the message for users)

Octachron replied:
I will have a look, because I think it is better to investigate that nothing deeper is going on before merging.

garrigue replied:
Sorry, I missed this answer and merged anyway (slightly tired of having to rebase several times a day).
Anyway, this is a corner case with printing that appears only once in the testsuite.
This said, it may reveal something strange in primitive aliases.

garrigue replied:
Looking better at the output, it looks more informative than before.
The type variable is probably caused by the aliasing code, and the `a` comes from de-aliasing. I'm not convinced that this use of aliasing is intuitive, but the cause seems to be on the primitive alias side, so if it is problematic, it should probably be fixed there.

smuenzel commented:
> I have included your 3 last commits, and reordered/squashed the commits so that we can merge keeping a reasonable number of separate commits. You can have a last look before I merge.

Looks good, thanks! I have no more concerns

garrigue commented:
Thanks to @smuenzel for a huge help in finishing this, @nojb for testing, and @gasche and @Octachron for trying to understand :)

Also, unfortunately we have not be able to create repro cases for most of the bugs found in opam with #14673, need to keep this in mind when looking at the code.
